functions/src/models/requests/timeline.ts

```diff
@@ -0,0 +1,96 @@
+import { FirestoreDataConverter } from '@google-cloud/firestore';
+import { IsEnum, IsObject } from 'class-validator';
+import { firestore } from 'firebase-admin';
+
+import { IOffer } from '../offers';
+import { IRequest, RequestStatus } from './index';
+import Timestamp = firestore.Timestamp;
+import DocumentData = firestore.DocumentData;
+
+export interface ITimelineItem extends DocumentData {
+  offerSnapshot: IOffer;
+  requestSnapshot: IRequest;
+  status: RequestStatus;
+  createdAt?: Timestamp;
+}
+
+/**
+ * Cannot be created/modified/deleted by a User. This is audit log maintained by the system
+ * It is used to render a timeline of the request on the front end, but all records
+ * are created automatically by the system based on status changes.
+ */
+export class TimelineItem implements ITimelineItem, FirestoreDataConverter<TimelineItem> {
+
+  @IsObject()
+  private _offerSnapshot: IOffer;
+
+  @IsObject()
+  private _requestSnapshot: IRequest;
+
+  @IsEnum(RequestStatus)
+  private _status: RequestStatus;
+
+  @IsObject()
+  private _createdAt: Timestamp;
+
+  constructor(offerSnapshot: IOffer, requestSnapshot: IRequest, status: RequestStatus, createdAt: Timestamp) {
+    this._offerSnapshot = offerSnapshot;
+    this._requestSnapshot = requestSnapshot;
+    this._status = status;
+    this._createdAt = createdAt;
```

Comment:
Do we have to assign now() for createdAt in the constructor itself, if the value is not supplied?

Reply:
Good point! Yes we should do this for the user.